internal/rbd/snap_diff.go

```diff
@@ -0,0 +1,263 @@
+	stream csi.SnapshotMetadata_GetMetadataAllocatedServer,
+	startingOffset int64,
+	maxResults int32) error {
+	image, err := rbdSnap.open()
+	if err != nil {
+		return err
+	}
+	defer image.Close()
+
+	snapID, err := rbdSnap.getRBDSnapID(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get snapshot ID of %q: %w", rbdSnap, err)
+	}
+
+	err = image.SetSnapByID(snapID)
+	if err != nil {
+		return fmt.Errorf("failed to set snapshot %q by ID %d: %w", rbdSnap, snapID, err)
+	}
+
+	changedBlocks := make([]*csi.BlockMetadata, 0)
+	diffIterateByIDConfig := librbd.DiffIterateByIDConfig{
+		Offset: uint64(startingOffset),
+		Length: uint64(rbdSnap.VolSize),
+		Callback: func(offset, sizeBytes uint64, _ int, _ interface{}) int {
+			select {
+			case <-stream.Context().Done():
+				return streamDoneErrCode
+			default:
+				changedBlocks = append(
+					changedBlocks,
+					&csi.BlockMetadata{
+						ByteOffset: int64(offset),
+						SizeBytes:  int64(sizeBytes),
+					})
+			}
+			if len(changedBlocks) < int(maxResults) {
+				// If we haven't reached the maxResults limit, continue collecting changed blocks.
+				return 0
+			}
+
+			// Send the current batch of changed blocks to the stream since
+			// we have reached the maxResults limit.
+			log.DebugLog(ctx, "Sending %d changed blocks to the stream", len(changedBlocks))
+			err := stream.Send(&csi.GetMetadataAllocatedResponse{
+				BlockMetadataType:   csi.BlockMetadataType_VARIABLE_LENGTH,
+				VolumeCapacityBytes: rbdSnap.VolSize,
+				BlockMetadata:       changedBlocks,
+			})
+			if err != nil {
+				log.ErrorLog(ctx, "failed to send response: %v", err)
+
+				return sendResponseErrCode
+			}
+			// Reset the changedBlocks slice for the next batch
+			changedBlocks = make([]*csi.BlockMetadata, 0)
+
+			return 0
+		},
+	}
+
+	err = image.DiffIterateByID(diffIterateByIDConfig)
+	if err != nil {
+		errno, ok := err.(interface{ ErrorCode() int })
+		if !ok {
+			// The error does not implement ErrorCode, return a generic internal error.
+			log.ErrorLog(ctx, "failed to get diff: %w", err)
+
+			return status.Error(codes.Internal, fmt.Sprintf("failed to get diff: $w", err))
+		}
+
+		switch errno.ErrorCode() {
+		case streamDoneErrCode:
+			// The stream was closed by the client.
+			// This is not an error, just a normal termination of the stream.
+			// Log the closure for debugging purposes.
+			log.DebugLog(ctx, "The stream was closed")
+
+			return nil
+		case sendResponseErrCode:
+			// An error occurred while sending the response.
+			log.ErrorLog(ctx, "failed to send response: %w", err)
+
+			return status.Error(codes.Internal, "failed to send response")
+		default:
+			// An unexpected error occurred.
+			log.ErrorLog(ctx, "failed to get diff: %w, error code: %d", err, errno)
+
+			return status.Error(codes.Internal, fmt.Sprintf("failed to get diff: %w, error code: %d", err, errno))
+		}
+	}
+	// If no error, we successfully completed the diff operation.
+	if len(changedBlocks) > 0 {
```

Comment:
Can we have a function to send the response and can be reused here and above like

```suggestion
send := func() error {
		if len(changedBlocks) == 0 {
			return nil
		}
		log.DebugLog(ctx, "Sending %d changed blocks to the stream", len(changedBlocks))
		if err := stream.Send(&csi.GetMetadataAllocatedResponse{
			BlockMetadataType:   csi.BlockMetadataType_VARIABLE_LENGTH,
			VolumeCapacityBytes: rbdSnap.VolSize,
			BlockMetadata:       changedBlocks,
		}); err != nil {
			log.ErrorLog(ctx, "failed to send response: %v", err)
			return status.Error(codes.Internal, fmt.Sprintf("failed to send response: %v", err))
		}
		changedBlocks = changedBlocks[:0] // Reset slice
		return nil
	}
```